File: src/896.单调数列.rs
/*
 * @lc app=leetcode.cn id=896 lang=rust
 *
 * [896] 单调数列
 */
struct Solution;
// @lc code=start
enum Status {
    Unknown,
    Ascending,
    Descending,
}
impl Solution {
    pub fn is_monotonic(a: Vec<i32>) -> bool {
        use std::cmp::Ordering;
        let mut status = Status::Unknown;
        for i in 0..a.len() - 1 {
            match status {
                Status::Unknown => {
                    status = match a[i].cmp(&a[i + 1]) {
                        Ordering::Less => Status::Ascending,
                        Ordering::Equal => Status::Unknown,
                        Ordering::Greater => Status::Descending,
                    }
                }
                Status::Ascending => {
                    if a[i] > a[i + 1] {
                        return false;
                    }
                }
                Status::Descending => {
                    if a[i] < a[i + 1] {
                        return false;
                    }
                }
            }
        }
        true
    }
}
// @lc code=end
#[test]
fn test_solution() {
    macro_rules! test {
        ($v:tt, $a:expr) => {
            assert_eq!(
                Solution::is_monotonic(vec!$v),
                $a
            );
        }
    };
    test!([1, 2, 2, 3], true);
    test!([6, 5, 4, 4], true);
    test!([1, 3, 2], false);
    test!([1, 2, 4, 5], true);
    test!([1, 1, 1], true);
}

File: src/896.单调数列.py
#
# @lc app=leetcode.cn id=896 lang=python3
#
# [896] 单调数列
#
from prelude import *
# @lc code=start
class Solution:
    def isMonotonic(self, A: List[int]) -> bool:
        # 0 for unknown, 1 for asc, -1 for desc
        status = 0
        for i in range(len(A) - 1):
            if status == 0:
                if A[i] == A[i+1]:
                    status = 0
                elif A[i] < A[i+1]:
                    status = 1
                else:
                    status = -1
            elif status == 1:
                if A[i] <= A[i+1]:
                    continue
                else:
                    return False
            else:
                if A[i] >= A[i+1]:
                    continue
                else:
                    return False
        return True

# @lc code=end

if __name__ == '__main__':
    def test(input, expected):
        calc = Solution().isMonotonic(input)
        if calc != expected:
            print(f'case failed: `{input}`')
            print(f'  calc     = `{calc}`')
            print(f'  expected = `{expected}`')
            exit(1)
    test([1,2,2,3], True)
    test([6,5,4,4], True)
    test([1,1,1], True)
    test([1,1,1,2,1], False)
